src/Core/GameState.flix

```diff
@@ -9,86 +10,124 @@ namespace Flixball/Core/GameState {
-        match (playerPositions0 |> Map.get(id)) {
-            // Case 1: No such player. Do nothing.
-            case None => state0
-
-            // Case 2: Found the player. Change their direction in the map and the board.
-            case Some(Position(coords, dir0)) => {
-                let dir = dir0 |> rotateDir(rot);
-                let playerPositions = playerPositions0 |> Map.insert(id, Position(coords, dir));
-                let board = (coords, board0) ||> adjust(tileOpt -> match tileOpt {
-                        case Some(Person(_, _)) => Some(Person(id, dir))
-                        case other => other
-                    }
-                );
-                GameState(states, moveLogics, playerPositions, board)
-            }
-        }
-    }
+    pub def rotatePlayer(id: PlayerId, rot: Rotation, gs: GameState): GameState =
+        gs |> adjust(id, info -> {
+            let pos = Position(posCoords(info.pos), posDir(info.pos) |> rotateDir(rot));
+            {pos = pos | info}
+        })
 
     /// Moves a player to a new position.
     /// Does not change the direction of the player.
-    pub def movePlayer(id: PlayerId, coords: Coordinates, state0: GameState): GameState = {
-        let GameState(states, moveLogics, playerPositions0, board0) = state0;
-        match (playerPositions0 |> Map.get(id)) {
-            // Case 1: No such player. Do nothing.
-            case None => state0
-
-            // Case 2: Found the player. Change their position in the map and the board.
-            case Some(Position(coords0, dir)) => {
-                let playerPositions = playerPositions0 |> Map.insert(id, Position(coords, dir));
-                match get_(coords0, board0) {
-                    case None => bug!("invalid game state")
-                    case Some(person) => {
-                        let board = board0 |> set_(coords0, None) |> set_(coords, Some(person));
-                        GameState(states, moveLogics, playerPositions, board)
-                    }
-                }
-            }
-        }
-    }
+    pub def movePlayer(id: PlayerId, coords: Coordinates, gs: GameState): GameState =
+        gs |> adjust(id, info -> {
+            let pos = Position(coords, posDir(info.pos));
+            {pos = pos | info}
+        })
     
     pub def mkGameState(seed: Int64, strategies: Map[PlayerId, Strategy], b: Board): GameState = {
-        let playerMap = Flixball/Core/Board.players(b) |> List.toMap;
-        let (states, moveLogics) = strategies |> Map.mapWithKey((id, strat) -> {
+        let playerMap = Flixball/Core/Board.players(b);
+        let ais = strategies |> Map.mapWithKey((id, strat) -> {
             let (stateGen, ml) = strat(id);
-            (stateGen(seed), ml)
-        }) |> Utils/Map.unzip;
-        GameState(states, moveLogics, playerMap, b)
+            let pos = Map.get(id, playerMap) |> Utils/Option.unsafeGet;
```

Comment:
I think every use of an `@Unsafe` function should have a comment explaining what invariant makes it safe.

Also consider: https://github.com/flix/flix/issues/4723